[#54377] Embedded work package attributes in PDF export

as-op opened the pull request:
https://community.openproject.org/work_packages/54377

- [x] Adapt pattern matching from HTML pipeline
    - [x] Test 
- [x] Implement replacement in export
    - [x] Test 
- [x] No replacement in markdown/html code and code blocks
    - [x] Test 
- [x] macros for rich text (e.g. description) are not supported
    - [x] Test 
- [x] Check for display rights of work package / project
    - [x] Test 
- [x] Check for display rights of project attributes
    - [x] Test

as-op committed:
[#54377] Embedded work package attributes in PDF export

https://community.openproject.org/work_packages/54377

as-op committed:
fix(lint): obey rubocop

as-op committed:
fix(export): pdf uses Yes/No for boolean custom fields; XLS & CVS do not

as-op committed:
extract I18n strings, adjust specs, test for excluded replacements
e.g. markdown/html code blocks

as-op committed:
test for permissions on attribute embeds

as-op committed:
check for display rights of project attributes

as-op committed:
test for display rights of project attributes

as-op committed:
fix missing variable for wp list to pdf report

ulferts commented on app/models/work_package/exports/macros/attributes.rb:
```diff
@@ -0,0 +1,185 @@
+      end
+
+      def self.resolve_value_project(project, attribute)
+        cf = CustomField.find_by(name: attribute, type: "ProjectCustomField")
```
This could also be written as 

`ProjectCustomField.find_by(name: attribute)`

ulferts commented on Gemfile:
```diff
@@ -157,9 +157,11 @@ gem "structured_warnings", "~> 0.4.0"
 # don't require by default, instead load on-demand when actually configured
 gem "airbrake", "~> 13.0.0", require: false
 
+gem "markly", "~> 0.10"
```
❓ I see that `markly` is added which does roughly the same thing CommonMarker does which we already have in the code base. 

I get why this is used here (took me a while to figure out) but it would be good to settle on one of the gems in the future.

ulferts replied:
A comment explaining where it is used would be nice.

as-op replied:
Commonmarker gem removed support fo the markdown abstract syntax tree some time ago.
I needed that for the attribute parsing in the pdf export, so  use the markly gem, which md_to_pdf already depends on it, too. So here just changed that it is a direct dependency now.

We could remove the commonmarker gem in the future and only use markly, but that was out of scope every time this topic came up.

I'll add a comment.

ulferts commented on app/models/exports/formatters/custom_field_pdf.rb:
```diff
@@ -0,0 +1,27 @@
+          object.typed_custom_value_for(custom_field)
+        else
+          object.formatted_custom_value_for(custom_field)
+        end
```
This looks like a broken inheritance to me. 

In `.apply` of super, the pdf format is disallowed and then the subclass changes it into the opposite by only allowing pdf. 

The `#format_for_export` method only changes a single line. 

Admittedly it might be formalism getting into my thinking as the only solution I would have is to have another superclass the two pdf and non pdf classes then inherit from.

ulferts commented on app/models/projects/exports/formatters/public.rb:
```diff
@@ -0,0 +1,42 @@
+      ##
+      # Takes a project and returns the localized status code
+      def format(project, **)
+        project.active ? I18n.t(:general_text_Yes) : I18n.t(:general_text_No)
```
I think this should be `project.public?` 

The comment seems to have been left unchanged after c&p

ulferts commented on app/models/projects/exports/formatters/active.rb:
```diff
@@ -0,0 +1,42 @@
+      ##
+      # Takes a project and returns the localized status code
+      def format(project, **)
+        project.active ? I18n.t(:general_text_Yes) : I18n.t(:general_text_No)
```
The comment seems to have been left unchanged after c&p

ulferts commented on app/models/work_package/exports/macros/attributes.rb:
```diff
@@ -0,0 +1,185 @@
+        content.include?("Label:") || content.include?("Value:")
+      end
+
+      def self.process_match(match, _matched_string, context)
```
The interface here could also be changed to named parameters instead of context which would save the deconstruction later on, so e.g. 

`def self.process_match(match, _matched_string, user:, work_package:)`

ulferts commented on app/models/work_package/exports/macros/attributes.rb:
```diff
@@ -0,0 +1,185 @@
+
+        work_package = WorkPackage.find_by(id:)
+        if work_package.nil? ||
+          !user.allowed_in_work_package?(:view_work_packages, work_package)
```
This could be simplified into

```
work_package = WorkPackage.visible(user).find_by(id:)
if work_package.nil?
```

ulferts commented on app/models/work_package/exports/macros/attributes.rb:
```diff
@@ -0,0 +1,185 @@
+        project = Project.find_by(id:)
+        if project.nil? ||
+          !user.allowed_in_project?(:view_project, project)
+          return msg_macro_error(I18n.t("export.macro.resource_not_found", resource: "#{Project.name} #{id}"))
```
The same simplification as for work_packages would be possible here.

ulferts commented on app/models/work_package/exports/macros/attributes.rb:
```diff
@@ -0,0 +1,185 @@
+          return msg_macro_error_rich_text if cf.formattable?
+
+          # Is the user allowed to see this custom field/"project attribute"?
+          if project.available_custom_fields.find { |pcf| pcf.id == cf.id }.nil?
```
If you are loading the project's available_custom_fields anyway, this could also be used to fetch it, right?

This would then change the method to:

```
return msg_macro_error_rich_text if DISABLED_PROJECT_RICH_TEXT_FIELDS.include?(ar_name.to_sym)

cf = project.available_custom_fields.find { |pcf| pcf.name == attribute }

return msg_macro_error_rich_text if cf&.formattable?

ar_name = if cf.nil?
            ::API::Utilities::PropertyNameConverter.to_ar_name(attribute.to_sym, context: project)
          else 
            "cf_#{cf.id}"
          end  

format_attribute_value(ar_name, Project, project)
```

There is a slight difference here compared to the other implementation which is that custom fields that exist but not in the project's context would more quickly fall back into the attempt of dealing with it via the static attributes. But I guess the problem of having a custom field which's name clashes with the built in attributes is an unsolved on in any case

ulferts commented on app/models/work_package/exports/macros/attributes.rb:
```diff
@@ -0,0 +1,185 @@
+
+        end
+
+        format_attribute_value(ar_name, WorkPackage, work_package)
```
This could be rewritten just like the method for Project. It would then also use the `work_package.available_custom_fields` which would answer the TODO.

ulferts replied:
If it turns out that both methods can be changed in the way proposed here, this probably also means that the code can be reused so maybe we end up only having a single method which just takes an additional parameter for either `WorkPackage` or `Project`.

ulferts commented on app/models/work_package/pdf_export/markdown_field.rb:
```diff
@@ -37,7 +38,52 @@ def write_markdown_field!(work_package, markdown, label)
+  end
+
+  def apply_macros(markdown, work_package, formatter)
+    return markdown unless formatter.applicable?(markdown)
```
The formatter in effect seems to be static since the method is only called from a single place if I am not mistaken. So it could then just become a local variable (or be placed into a method)

ulferts commented on app/models/work_package/pdf_export/markdown_field.rb:
```diff
@@ -37,7 +38,52 @@ def write_markdown_field!(work_package, markdown, label)
+  end
+
+  def apply_macro_text(text, work_package, formatter)
+    return text unless formatter.applicable?(text)
```
This has already been checked in the calling method and this method here is private. Admittedly it is in a module so an including class could change this but currently that is not the case and somewhat unlikely I think.

as-op replied:
It's also called in apply_macro_html_node, so this check is useful to skip costly html parsing.

In apply_macro_html() it is useless, indeed.

ulferts commented on app/models/work_package/pdf_export/markdown_field.rb:
```diff
@@ -37,7 +38,52 @@ def write_markdown_field!(work_package, markdown, label)
+  end
+
+  def apply_macro_html(html, work_package, formatter)
+    return html unless formatter.applicable?(html)
```
Same as above, this has been checked before in the callstack.

ulferts commented on app/models/work_package/exports/macros/attributes.rb:
```diff
@@ -0,0 +1,185 @@
+        else
+          ar_name = "cf_#{cf.id}"
+          # currently we do not support embedding rich text fields: long text custom fields
+          return msg_macro_error_rich_text if cf.formattable?
```
If the field is formattable but disabled in the project, the error returned will still be

`[Rich text embedding currently not supported in export]`

when one would expect to get a message that the 'resource not found' error.

ulferts commented on app/models/work_package/exports/macros/attributes.rb:
```diff
@@ -0,0 +1,185 @@
+    #   workPackageValue:1234:subject # Outputs the actual subject of #1234
+    #
+    #   projectLabel:statusExplanation # Outputs current project label attribute "Status description" + help text
+    #   projectValue:statusExplanation # Outputs current project value for "Status description"
```
The macro for the project also supports a [reference to a different project](https://www.openproject.org/docs/user-guide/wysiwyg/#available-attributes-for-projects). Currently, the PDF export does not support it.

ulferts reviewed (changes requested):
Hey @as-op please find my remarks next to the code. 

Admittedly I skipped the specs since I wanted to provide feedback more quickly.

as-op committed:
use ProjectCustomField.find_by

https://github.com/opf/openproject/pull/15702#discussion_r1627286353

as-op committed:
add markly comment

https://github.com/opf/openproject/pull/15702#discussion_r1627296963

as-op committed:
fix typo

as-op committed:
fix bug and comment in attribute formatter project.public

as-op committed:
fix comment in attribute formatter project.active

as-op committed:
simplify wp access check & find
https://github.com/opf/openproject/pull/15702#discussion_r1627409100

as-op committed:
simplify project access check & find

https://github.com/opf/openproject/pull/15702#discussion_r1627409660

as-op committed:
simplify & unify attribute value resolving

https://github.com/opf/openproject/pull/15702#discussion_r1627436214
https://github.com/opf/openproject/pull/15702#discussion_r1627444854

as-op committed:
remove already checked formatter.applicable?

https://github.com/opf/openproject/pull/15702#discussion_r1627479843

as-op committed:
support referencing project by identifier

https://github.com/opf/openproject/pull/15702#discussion_r1627668679

as-op committed:
list more examples in comment
